src/Common/ProgressBar.cpp

```diff
@@ -0,0 +1,111 @@
+#include "ProgressBar.h"
+#include <IO/WriteBufferFromFileDescriptor.h>
+#include <Common/TerminalSize.h>
+#include <Common/UnicodeBar.h>
+#include <Databases/DatabaseMemory.h>
+
+namespace DB
+{
+
+bool ProgressBar::updateProgress(Progress & progress, const Progress & value)
+{
+    return progress.incrementPiecewiseAtomically(value);
+}
+
+void ProgressBar::writeProgress(const Progress & progress, const size_t elapsed_ns)
+{
+    if (!need_render_progress)
+        return;
+
+    /// Output all progress bar commands to stderr at once to avoid flicker.
+    WriteBufferFromFileDescriptor message(STDERR_FILENO, 1024);
+
+    static size_t increment = 0;
+    static const char * indicators[8] = {
```

Comment:
`static constexpr const char* indicators[8]` is better, you can also use `static constexpr std::string_view []` to avoid calling strlen for output streams